### PR TITLE
Ks 296 dispatcher context and threading

### PR DESCRIPTION
**Pull request by ettec**
The threading model chosen is one thread per capability, reasoning for this is as follows:

1) Prevents a 'dodgy' capability impacting others within the node by blocking the receive thread
2) Makes it easy to monitor per capability performance by monitoring the receive queue size

3) This third benefit is debatable, small and something we may want to revisit once we understand with experience how load should be balanced across capabilities.....having said that....initally at least it is advantageous to have the Receive method of the capabilities single threaded as it keeps things relatively simple and easier to reason about should we encounter any issues.

jira issue: https://smartcontract-it.atlassian.net/browse/KS-296

https://github.com/smartcontractkit/chainlink/pull/13524. depends on this PR.

### CODE REVIEW

File: core/capabilities/remote/target/endtoend_test.go
Comment (bolekk) [inline core/capabilities/remote/target/endtoend_test.go]: We usually use tests.Context(t) in tests.

File: core/capabilities/remote/dispatcher.go
Comment (bolekk) [inline core/capabilities/remote/dispatcher.go:80]: Maybe prefix with "dispatcher_" instead of "capability_"?

File: core/capabilities/remote/dispatcher.go
Comment (bolekk) [inline core/capabilities/remote/dispatcher.go]: I think it's fine - capability should be responsible here.

File: core/capabilities/remote/dispatcher.go
Comment (bolekk) [inline core/capabilities/remote/dispatcher.go]: I'd rather spam logs than drop silently. We can deal with log rate limiting later.

File: core/capabilities/remote/dispatcher.go
Comment (bolekk) [inline core/capabilities/remote/dispatcher.go]: Yeah, sounds good to me.

File: core/capabilities/remote/target/client.go
Comment (bolekk) [inline core/capabilities/remote/target/client.go]: Remove these comments before merging.

**Review by bolekk — COMMENTED** [review]
I like the approach, thanks!

File: core/capabilities/remote/dispatcher.go
Comment (ettec) [reply core/capabilities/remote/dispatcher.go:80]: its the usage of the capabilities receive channel, so i this capability is the best name

File: core/capabilities/remote/target/client.go
Comment (ettec) [reply core/capabilities/remote/target/client.go]: will do, going to leave comments in place until I have the second review in place

File: core/capabilities/remote/dispatcher.go
Comment (ettec) [reply core/capabilities/remote/dispatcher.go]: fair enough - I've add that in - left the comments for now until second review done

Comment (cedric-cordenier) [reply core/capabilities/remote/dispatcher.go]: @ettec Why drop it immediately vs waiting some minimal amount of time?

**Review by cedric-cordenier — DISMISSED** [review]

File: core/capabilities/remote/dispatcher.go
Comment (ettec) [reply core/capabilities/remote/dispatcher.go]: if its being dropped, that implies the queue is full,  which has a buffer size of 10k, so its safe to assume if it gets to 10k queue backlog the capability is hosed and theres no point waiting

Comment (ettec) [reply core/capabilities/remote/dispatcher.go]: the queue monitoring will highlight if there are issues way before that point in practice

**Review by krehermann — APPROVED** [review]

**Review by bolekk — APPROVED** [review]

**Pull request merged**
